// @videogular/ngx-videogular/esm2020/overlay-play/lib/overlay-play.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VgOverlayPlayComponent } from './vg-overlay-play.component';
import { VgCoreModule } from '@videogular/ngx-videogular/core';
import * as i0 from "@angular/core";
export class VgOverlayPlayModule {
}
/** @nocollapse */ VgOverlayPlayModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: VgOverlayPlayModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ VgOverlayPlayModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.6", ngImport: i0, type: VgOverlayPlayModule, declarations: [VgOverlayPlayComponent], imports: [CommonModule, VgCoreModule], exports: [VgOverlayPlayComponent] });
/** @nocollapse */ VgOverlayPlayModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: VgOverlayPlayModule, imports: [CommonModule, VgCoreModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: VgOverlayPlayModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, VgCoreModule],
                    declarations: [VgOverlayPlayComponent],
                    exports: [VgOverlayPlayComponent],
                }]
        }] });
