// @videogular/ngx-videogular/esm2020/modulo/lib/modulo.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { VgModuloComponent } from "./components/vg-modulo.component";
import * as i0 from "@angular/core";
export class VgModuloModule {
}
/** @nocollapse */ VgModuloModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: VgModuloModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ VgModuloModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.6", ngImport: i0, type: VgModuloModule, declarations: [VgModuloComponent], imports: [CommonModule], exports: [VgModuloComponent] });
/** @nocollapse */ VgModuloModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: VgModuloModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: VgModuloModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [VgModuloComponent],
                    exports: [VgModuloComponent],
                }]
        }] });
