// @videogular/ngx-videogular/esm2020/streaming/lib/streaming.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VgDashDirective } from './directives/vg-dash/vg-dash.directive';
import { VgHlsDirective } from './directives/vg-hls/vg-hls.directive';
import { VgCoreModule } from '@videogular/ngx-videogular/core';
import * as i0 from "@angular/core";
export class VgStreamingModule {
}
/** @nocollapse */ VgStreamingModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: VgStreamingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ VgStreamingModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.6", ngImport: i0, type: VgStreamingModule, declarations: [VgDashDirective, VgHlsDirective], imports: [CommonModule, VgCoreModule], exports: [VgDashDirective, VgHlsDirective] });
/** @nocollapse */ VgStreamingModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: VgStreamingModule, imports: [CommonModule, VgCoreModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: VgStreamingModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, VgCoreModule],
                    declarations: [VgDashDirective, VgHlsDirective],
                    exports: [VgDashDirective, VgHlsDirective]
                }]
        }] });
