// @videogular/ngx-videogular/esm2020/ima-ads/lib/ima-ads.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VgImaAdsComponent } from './vg-ima-ads.component';
import { VgCoreModule } from '@videogular/ngx-videogular/core';
import * as i0 from "@angular/core";
export class VgImaAdsModule {
}
/** @nocollapse */ VgImaAdsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: VgImaAdsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ VgImaAdsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.6", ngImport: i0, type: VgImaAdsModule, declarations: [VgImaAdsComponent], imports: [CommonModule, VgCoreModule], exports: [VgImaAdsComponent] });
/** @nocollapse */ VgImaAdsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: VgImaAdsModule, imports: [CommonModule, VgCoreModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: VgImaAdsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, VgCoreModule],
                    declarations: [VgImaAdsComponent],
                    exports: [VgImaAdsComponent],
                }]
        }] });
